spec.html

```diff
@@ -241,11 +244,36 @@ <h1>Static Semantics: HostGetSupportedImportAssertions ()</h1>
     <emu-note type=editor>The purpose of requiring the host to specify its supported import assertions, rather than passing all assertions to the host and letting it then choose which ones it wants to handle, is to ensure that unsupported assertions are handled in a consistent way across different hosts.</emu-note>
   </emu-clause>
 
+  <emu-clause id="sec-hostgetsupportedextramodulescripttypes" aoid="HostGetSupportedExtraModuleScriptTypes">
+    <h1>Static Semantics: HostGetSupportedExtraModuleScriptTypes ()</h1>
+    <p>
+      HostGetSupportedExtraModuleScriptTypes is a host-defined abstract operation that allows host environments to specify which types of module scripts they support in addition to Source Text Module Record.
+      Module script type assertions with unsupported values will cause an error.
+    </p>
+
+    <p>The implementation of HostGetSupportedExtraModuleScriptTypes must conform to the following requrements:</p>
+
+    <ul>
+      <li>It must return a List whose values are all StringValues, each indicating a supported module script type.</li>
+
+      <li>None of the string values may be *"js"* or *"javascript"*.<emu-note type="editor">It is assumed that Source Text Module Record must be supported.</emu-note></li>
+
+      <li>Each time this operation is called, it must return the same List instance with the same contents.</li>
```

Comment:
Per-realm, or program-wide?

Reply:
I assume we want program-wide here.  I would be surprised if a host would change the module types it supports on a per-realm basis. I could be wrong about that though.

I was mimicking language I saw for things like [HostHasSourceTextAvailable](https://tc39.es/ecma262/#sec-hosthassourcetextavailable): "Each time it is called with a specific func as its argument, it must return the same completion record.".  I guess that probably means per-realm, though?

Is there a formal way to express that something must return the same thing program-wide -- assuming that's what we actually want here?

Reply:
I'm not actually certain here. It might be that every AO is already implicitly per-realm, which would mean this wording doesn't force it across every realm.

cc @erights @caridy @leobalter for any relation to the Realms proposal, cc @syg @bakkot @michaelficarra for the question as it pertains to the actual spec.

Reply:
for Realms, we are proposing that each Realm has a different module map, but I assume the module types mapping could be just reused. In this case, I believe it wouldn't be an issue. This mapping won't share Realm identities or any fingerprints, anyway.

I believe @erights might have a different perspective for the compartments proposal, but I can't answer for that without more context.

Reply:
This has to at least be per agent. We'd definitely not support HTML modules in workers or worklets for instance.